### ISSUE TITLE
Follow damp jitter at <60 FPS

### ISSUE THREAD

**Issue by WeaselOnaStick** [ramokz/phantom-camera#206]
### Issue description

The video should explain it pretty well. Basically if the camera is set to follow a rigid body, and the follow has damping, then if FPS dips below 60 you will get this nasty jitter. I added a simple overhead camera to verify that the actual position of said rigidbody isn't jittering itself. Tested with all camera follow modes with/without damping (even toggled between jolt/godot's physics) and I can confirm that it's because of damping.

https://github.com/ramokz/phantom-camera/assets/10953814/3ccae39f-111d-41bd-bb7a-5394b7822d4e

### Steps to reproduce

Simple vehicle
Phantom Camera with follow + damping

### (Optional) Minimal reproduction project

[dispensable.zip](https://github.com/ramokz/phantom-camera/files/14137188/dispensable.zip)

**Comment by ramokz** [comment]
Thanks for reporting this and sharing the project files!

Am doing a refactor of damping system in the near future (#92).
Will validate the changes with your attached project and see if that solves it.

**Comment by TCROC** [comment]
I managed to solve this on my project today! Good news is: it didn't require making any changes to phantom-camera! I'm on mobile right now and will post a more detailed summary tomorrow as the problem is a tricky one to wrap your head around but is actually quite simple to solve.

There are a few things that can cause this problem:

# 1. Different update rates

I suspect this is the exact issue you are encountering above. The solution is to make sure your camera and physics object update their positions at the same rate. When you reduce the frame rate, your camera is updating its position at a slower rate than the physics object because the physics object updates independent of your frame rate. Physics engines tend to run at either 50 or 60 steps per second. In your cause probably 60 since 60 appeared smooth for you.

How to solve?

3 approaches:

1. Set the process more on the camera to disabled and manually call the camera's _process method from another scripts _physics_process method. This will ensure the physics and camera move in sync with each other. The trade off is that your surrounding environment will likely appear jumpy. It's less noticeable, but we can do better ;)

2. Set your physics update rate to match your fps. Then your camera and physics objects are again updating in sync! :) ... DONT DO THIS! This will cause your physics engine to do strange things as the frame rate fluctuates. Especially low FPS. You will see your vehicle turn into a rocket, launch to Mars, and turn into the latest and greatest Mars rover ;)

3. Interpolate the 3d physics nodes yourself between fixed updates! This is the one I use! :) Basically, you cache the previous point in 3d space the physics object was at, its current point, and then smoothly interpolate to those points in the _process method between _physics_process steps. As always, there are some trade offs to this approach. 

The physics body will always be rendering the previous physics step so will always be in the past. It may be prone to clipping (very very small) and prone to not visually showing exact contact (again very very small). So small that it's unnoticeable in fast paced games.

It's what we use in our game over here: https://store.steampowered.com/app/1343040/Blocky_Ball/

We are currently porting it from Unity to Godot. This asset is proving to be a tremendous Cinemachine alternative! 

I'll grab some more details for you tomorrow and share the resources I studied. And if I get some time later, I'll take a look at your example project and see if I can fix the stutter. Then we'll have a nice example to share with the community if you want to contribute it! :) 

(Typed all this on my phone. Hopefully autocomplete did its job well :) )

**Comment by TCROC** [comment]
Oh and I realized I only talked about point 1 that I have found to cause jitter (which I believe to be this issue). But I'll list them all out for clarity tomorrow as well

**Comment by williamviktorsson** [comment]
I had quite som jitter as well on both my Mac and PC which went away completely when i swapped the `_process(delta: float)` to a `_physics_process(delta: float)` in the `phantom_camera_2D.gd` script.

**Comment by TCROC** [comment]
This is correct.  I am very curious to see which mitigation techniques are used in 0.7! :)  Between Unity's Cinemachine and now this amazing addon, I have not yet found a general purpose solution to the jitter.  Even in my own custom projects.  The only solution I've been able to find to consistently work and improve the experience is in summary:

Update and move the camera at the same rate as the objects it is focused on.  Then move the surrounding objects at the same rate as well.  But the most important one is the objects in focus as that is where the player's eyes will be drawn to.

Sorry I haven't posted my example project.  Had some personal things in life come up that detracted away.  I'll get around to it eventually as I think it'll be helpful for learning and a good point for us to collectively improve upon.

**Comment by Xorblax** [comment]
I think an option to use cubic hermite interpolation as opposed to linear would solve this problem for a lot of people.

**Comment by ramokz** [comment]
@Xorblax Could you elaborate on how that interpolation method works, and why it would work better?

The current implementation doesn't use a linear interpolation, but rather a [smoothdamp](https://github.com/ramokz/phantom-camera/blob/727d42558470d826eeb8c0b3eef0d3f2b78ef04b/addons/phantom_camera/scripts/phantom_camera/phantom_camera_3d.gd#L1044) based on [Unity's approach](https://github.com/Unity-Technologies/UnityCsReference/blob/02d565cf3dd0f6b15069ba976064c75dc2705b08/Runtime/Export/Math/Mathf.cs#L301), which in turn is based on the [critical damping](https://www.alexisbacot.com/blog/the-art-of-damping) equation.

**Comment by Reuzi3** [comment]
any way to get around this? I’m experiencing jitter when I enable damping- 0.1 isn’t very noticeable, but higher values cause a lot of shaking when camera try stop. When I disable it everything works perfectly. I’ve tried several damping methods and none of them work; the only thing that helps is snapping the camera and rounding the position, but that keeps the damping from being smooth and slow as it approaches the player’s position

**Comment by ramokz** [comment]
@Reuzi3 Would you be able to share a [MRP](https://blog.3d-logic.com/2024/05/09/why-should-you-care-about-minimal-reproducible-examples-and-how-to-create-one/)?

Aside from cases that involve built-in Godot physics nodes, like the one in this issue, I haven't come across, or seen any reports, about jitter in other scenarios ever since physics interpolation support was added.

**Comment by Reuzi3** [comment]
> [@Reuzi3](https://github.com/Reuzi3) Would you be able to share a [MRP](https://blog.3d-logic.com/2024/05/09/why-should-you-care-about-minimal-reproducible-examples-and-how-to-create-one/)?
> 
> Aside from cases that involve built-in Godot physics nodes, like the one in this issue, I haven't come across, or seen any reports, about jitter in other scenarios ever since physics interpolation support was added.

Here it! my monitor runs at 144 Hz.. The jitter gets more noticeable the more assets you have in the scene, and when it’s about to stop

https://github.com/user-attachments/assets/b8ee3537-a037-47db-8508-b5d576166cce

[jitter.ZIP](https://github.com/user-attachments/files/20218311/jitter.ZIP)

**Comment by ramokz** [comment]
@Reuzi3 Would you be able to make a [new issue](https://github.com/ramokz/phantom-camera/issues/new/choose) or a [discussion post](https://github.com/ramokz/phantom-camera/discussions/new/choose) for this?

Have some notes for what you shared, however, the thing you're seeing isn't related to the issue in this post.
You're having an issue with a 2D node scene, whereas the issue in this post is about jitter in a 3D scene that uses built-in physics-based nodes, which is quite a different context.

Basically, whenever someone writes a comment on an existing [issue](https://github.com/ramokz/phantom-camera/issues/new/choose), or discussion post, everyone who has previously engaged in the issue will get a notification and often also an email. But, again, the thing you're referring to isn't related to the issue described in this post, folks will, in their eyes, get spam notifications. Along with making the conversation of the initial issue harder to keep track of.
